hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/HddsVolume.java

```diff
@@ -214,28 +223,32 @@ public void loadDbStore() throws IOException {
     }
 
     dbParentDir = storageIdDir;
+    dbLoaded.set(true);
+    LOG.info("SchemaV3 db is loaded at {} for volume {}", containerDBPath,
+        getStorageID());
   }
 
   /**
    * Pick a DbVolume for HddsVolume and init db instance.
    * Use the HddsVolume directly if no DbVolume found.
    * @param dbVolumeSet
    */
-  public void createDbStore(MutableVolumeSet dbVolumeSet)
-      throws IOException {
+  public void createDbStore(MutableVolumeSet dbVolumeSet) throws IOException {
     DbVolume chosenDbVolume = null;
     File clusterIdDir;
+    String workingDir = getWorkingDir() == null ? getClusterID() :
```

Comment:
This is good, just to confirm my understanding, the newly added `workingDir` is assigned a value only during format, so `getWorkingDir() == null` means we are doing a `finalizeUpgrade` operation otherwise we are formatting a volume, right?

I've got an idea, if `createDbStore` rely on `workingDir` for now instead of clusterID, then maybe we should use `workingDir` in `loadDbStore` too?
For example we could scan for the on-disk working dir when we do `initialize()` for a volume and assign a value to `workingDir` if it exists?

Reply:
You are right.  workingDir != null means we have formatted the volume this time DN starts, otherwise vice versa. 

loadDbStore will be called only when schema V3 is finalized. So, workingDir will definitely be the clusterID. There is no doubt of that.

Reply:
There is a case that SchemaV3 is enabled but SCMHA is not enabled(even they are all finalized), then the working dir is actually the ScmID instead of the ClusterID.

Reply:
In this case, we need persistent the DB path.

Reply:
Oh, I rechecked the code here: `StorageVolumeUtil.checkVolume()`, I think the case provided by me is not good.
With the current logic, a link from "ScmIdDir -> ClusterIdDir" will automatically be created with SCMHA finalized within `VersionedDatanodeFeatures.ScmHA.upgradeVolumeIfNeeded`, so the working dir will always be the "ClusterIdDir" indeed. Sorry for the noise.